src/safeposix/syscalls/fs_calls.rs

```diff
@@ -1537,4 +1537,80 @@ impl Cage {
+            let mut filedesc_enum = wrappedfd.write().unwrap();
+            
+            match &mut *filedesc_enum {
+                // only proceed when fd represents a file
+                File(ref mut normalfile_filedesc_obj) => {
+                    let metadata = FS_METADATA.read().unwrap();
+                    let inodeobj = metadata.inodetable.get(&normalfile_filedesc_obj.inode).unwrap();
+
+                    match inodeobj {
+                        // only proceed when inode is a dir
+                        Inode::Dir(dir_inode_obj) => {
+                            let position = normalfile_filedesc_obj.position;
+                            let mut bufcount = 0;
+                            let mut curr_size = 0;
+                            let mut count = 0;
+                            let mut temp_len = 0;
+
+                            // iterate over filename-inode pairs in dict
+                            for (filename, inode) in dir_inode_obj.filename_to_inode_dict.clone().into_iter() {
+                                // convert filename to a filename vector of u8
+                                let mut vec_filename: Vec<u8> = filename.as_bytes().to_vec();
+                                
+                                vec_filename.push(DT_UNKNOWN); // push DT_UNKNOWN as d_type (for now)
+                                temp_len = vec_filename.len(); // get length of current filename vector for padding calculation
+                                
+                                // pad filename vector to the next highest 8 byte boundary
+                                for n in 0..(temp_len + 7) / 8 * 8 - temp_len {
+                                    vec_filename.push(00);
+                                }
+                                
+                                // the fixed dirent size and length of filename vector add up to total size
+                                curr_size = interface::CLIPPED_DIRENT_SIZE + vec_filename.len(); 
+                                
+                                bufcount += curr_size; // increment bufcount
+                                if bufcount > bufsize {break;} // stop iteration if current bufcount exceeds argument bufsize
+                                
+                                // push properly constructed tuple to vector storing result
+                                vec.push((interface::ClippedDirent{d_ino: inode as u64, d_off: bufcount as u64, d_reclen: curr_size as u16}, vec_filename));
+                                count += 1;
+                            }
+                            // update file position
+                            normalfile_filedesc_obj.position = interface::rust_min(position + count, dir_inode_obj.filename_to_inode_dict.len());
+                            
+                            interface::pack_dirents(vec, dirp);
+
+                            0 // getdents succeeded
```

Comment:
The number of bytes written is actually what is returned, not 0,  so you can get that from pack_dirents by subtracting baseptr from the last curptr (using a wrapping_sub probably so as not to need unsafe code) and then casting that to a usize then an i32.